application/apps/indexer/sources/src/serial/serialport.rs

```diff
@@ -0,0 +1,131 @@
+use crate::ByteSource;
+use crate::{Error as SourceError, ReloadInfo, SourceFilter};
+use async_trait::async_trait;
+use buf_redux::Buffer;
+use std::time::Duration;
+use tokio::io::AsyncReadExt;
+use tokio_serial::{SerialPortBuilderExt, SerialStream};
+
+const TIMEOUT: u64 = 100;
+
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum SerialError {
+    #[error("Serial setup problem: {0}")]
+    Setup(String),
+    #[error("Unrecoverable serial error: {0}")]
+    Unrecoverable(String),
+}
+
+pub struct SerialSource {
+    port: SerialStream,
+    buffer: Buffer,
+    read: [u8; 1],
+    amount: usize,
+}
+
+impl SerialSource {
+    pub fn new(path: &str, baud_rate: u32) -> Result<Self, SerialError> {
+        Ok(Self {
+            buffer: Buffer::new(),
+            read: [0; 1],
+            amount: 0,
+            port: tokio_serial::new(path, baud_rate)
+                .open_native_async()
+                .map_err(|e| SerialError::Unrecoverable(format!("Could not open port: {}", e)))?,
+        })
+    }
+}
+
+#[async_trait]
+impl ByteSource for SerialSource {
+    async fn reload(
+        &mut self,
+        _filter: Option<&SourceFilter>,
+    ) -> Result<Option<ReloadInfo>, SourceError> {
+        let mut copied: usize = 0;
+        let mut buffer_error: String = "".to_string();
+        let timeout = tokio::time::timeout(Duration::from_millis(TIMEOUT), async {
+            loop {
+                if let Err(err) = self.port.read_exact(&mut self.read).await {
+                    buffer_error = format!("read from serial stream failed: {}", err);
+                };
+                copied += self.buffer.copy_from_slice(&self.read);
+            }
+        })
+        .await;
+        if timeout.is_err() {
+            Ok(Some(ReloadInfo::new(0, 0, 0, None)))
+        } else if buffer_error.is_empty() {
+            self.amount = copied;
+            Ok(Some(ReloadInfo::new(self.amount, self.amount, 0, None)))
+        } else {
+            Err(SourceError::Unrecoverable(buffer_error))
+        }
+    }
+
+    fn current_slice(&self) -> &[u8] {
+        self.buffer.buf()
+    }
+
+    fn consume(&mut self, offset: usize) {
+        self.buffer.consume(offset);
+    }
+
+    fn len(&self) -> usize {
+        self.buffer.len()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+#[tokio::test]
+async fn test_serial() {
```

Comment:
if not used please remove or at least comment out...otherwise a successful test is reported